### PR TITLE
Added test for regression on module import

### PR DESCRIPTION
**Pull request by abhishekpratapa**

### CODE REVIEW

File: src/python/turicreate/test/test_modules.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+# Use of this source code is governed by a BSD-3-clause license that can
+# be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+from __future__ import print_function as _
+from __future__ import division as _
+from __future__ import absolute_import as _
+import unittest
+import turicreate as tc
+import inspect
+
+class LoadModuleTest(unittest.TestCase):
```
Comment (srikris) [inline src/python/turicreate/test/test_modules.py]: I think we have a test_api_visilibiy. You should add this as an additional test there or make that test work :)

Comment (abhishekpratapa) [reply src/python/turicreate/test/test_modules.py]: K!

File: src/python/turicreate/test/_test_api_visibility.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 from __future__ import absolute_import as _
 import unittest
 import turicreate
```
Comment (srikris) [inline src/python/turicreate/test/_test_api_visibility.py]: This is an _ test. Does it actually run?

Comment (abhishekpratapa) [reply src/python/turicreate/test/_test_api_visibility.py]: Not by default. It does when we run:

```bash
 pytest _test_api_visibility.py 
```

Comment (abhishekpratapa) [reply src/python/turicreate/test/_test_api_visibility.py]: Most of the tests in here fail though.

Comment (srikris) [reply src/python/turicreate/test/_test_api_visibility.py]: Can we make it run by default. Maybe a good time to actually consider doing it.

Comment (abhishekpratapa) [reply src/python/turicreate/test/_test_api_visibility.py]: I can update that file if we want those tests to pass. We just haven't been updating it every time we add new api.

**Comment by abhishekpratapa** [comment]
Tests to be fixed 
- [x]  TuriTests.test_top_level
- [ ] TabCompleteVisibilityTests.test_churn_predictor
- [ ] TabCompleteVisibilityTests.test_kmeans
- [ ] TabCompleteVisibilityTests.test_local_outlier_factor
- [ ] TabCompleteVisibilityTests.test_search
- [ ] TabCompleteVisibilityTests.test_supervised
- [ ] TabCompleteVisibilityTests.test_topic_model
- [ ] ModuleVisibilityTests.test_Image_type
- [ ] ModuleVisibilityTests.test_anomaly_detection
- [ ] ModuleVisibilityTests.test_churn_predictor
- [ ] ModuleVisibilityTests.test_classifier
- [ ]  ModuleVisibilityTests.test_cross_validation
- [ ] ModuleVisibilityTests.test_data_matching
- [ ] ModuleVisibilityTests.test_feature_engineering
- [ ] ModuleVisibilityTests.test_graph_analytics
- [ ] ModuleVisibilityTests.test_lead_scoring
- [ ] ModuleVisibilityTests.test_models_with_hyper_parameters
- [ ] ModuleVisibilityTests.test_nearest_neighbors
- [ ] ModuleVisibilityTests.test_recommender
- [ ] ModuleVisibilityTests.test_text_analytics
- [ ] ModuleVisibilityTests.test_toolkits
- [ ] ModuleVisibilityTests.test_topic_model
- [ ] ModuleVisibilityTests.test_topic_modelling_models

File: src/python/turicreate/test/_test_api_visibility.py (hunk 2 above)
Comment (abhishekpratapa) [reply src/python/turicreate/test/_test_api_visibility.py]: Done

File: src/python/turicreate/test/test_api_visibility.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+                    'boosted_trees_regression',
+                    'decision_tree_regression',
+                    'random_forest_regression']
+
+        check_visible_modules(actual, expected)
+
+    def test_toolkits(self):
+        actual = get_visible_items(turicreate.toolkits)
+
+        ## TODO: LazyModule has hidden some of these imports for us. We need to
```
Comment (srikris) [inline src/python/turicreate/test/test_api_visibility.py]: This is by design. You might want to remove this todo :)

File: src/python/turicreate/test/test_api_visibility.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+        ##       fix this somehow. Works perfectly in the Repel.
+        
+        expected = [
+                    "classifier",
+                    "clustering",
+                    "distances",
+                    "evaluation",
+                    "graph_analytics",
```
Comment (srikris) [inline src/python/turicreate/test/test_api_visibility.py]: Shouldn't there be more modules visible. Are you sure this is actually testing the right thing.

**Comment by Jarvi-Izana** [comment]
Great job. The python import is such a tricky part. Let me find out who introduced config.

for other missing pieces:
run  `ipython3 -c "import turicreate as tc; dir(tc)" >> foo.txt` on both pypi turicreate and local turicreate with lazy import:
```
>  '_LazyCallable',
>  '_LazyModuleLoader',
31a34
>  'audio_analysis',
36d38
<  'config',
87c89
<  'topic_model',
---
>  'topci_model',
```
you catch the topic model, still I forget audio_analysis and config.

**Comment by Jarvi-Izana** [comment]
minor fix: you need to add `config` and possibly remove `audio_analysis` since it's not in 5.6. Surprisingly, audio_analysis is not in our public api and does it mean no user?

File: src/python/turicreate/test/test_api_visibility.py (hunk 4 above)
Comment (abhishekpratapa) [reply src/python/turicreate/test/test_api_visibility.py]: Yeah, they should be. The `TODO` was more of a mental note for myself that I needed to add more toolkits to this test. I can add them now.

File: src/python/turicreate/test/test_api_visibility.py (hunk 3 above)
Comment (abhishekpratapa) [reply src/python/turicreate/test/test_api_visibility.py]: Solid. See comment below. Working on it!

**Comment by abhishekpratapa** [comment]
@Jarvi-Izana https://github.com/apple/turicreate/pull/2352

Fixes the config import. Not sure about `audio_analysis`, if that was something you introduced accidentally or is actually an expected feature.

**Comment by Jarvi-Izana** [comment]
the lazy module loader will hide the import of tc.config since all the toolkit imports below are deferred.
```
src/unity/python/turicreate/toolkits/image_classifier/image_classifier.py
22:from turicreate import config as _tc_config

src/unity/python/turicreate/toolkits/object_detector/object_detector.py
28:from turicreate import config as _tc_config

src/python/turicreate/util/__init__.py
30:    import configparser as _ConfigParser
32:    import ConfigParser as _ConfigParser

src/python/turicreate/toolkits/image_classifier/image_classifier.py
22:from turicreate import config as _tc_config

src/python/turicreate/toolkits/object_detector/object_detector.py
28:from turicreate import config as _tc_config
```
let's make those implicit imports as explicit imports, as python zen suggested.

File: src/python/turicreate/__init__.py [hunk 5]
```diff
@@ -90,7 +90,7 @@
 # we can use top-level lazy import for them
 distances = _LazyModuleLoader('turicreate.toolkits.distances')
 nearest_neighbors = _LazyModuleLoader('turicreate.toolkits.nearest_neighbors')
-topci_model = _LazyModuleLoader('turicreate.toolkits.topci_model')
+topic_model = _LazyModuleLoader('turicreate.toolkits.topic_model')
```
Comment (Jarvi-Izana) [inline src/python/turicreate/__init__.py]: Hmmm, I'm surprised no test is for this module and I even passed the smoke test. Probably this module is no longer in use?

File: src/python/turicreate/test/test_api_visibility.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+
+class TuriTests(unittest.TestCase):
+    def test_top_level(self):
+        for x in (MODULES + CLUSTER + DATA_STRUCTURES + GRAPH_ANALYTICS +
+                  GENERAL + RECOMMENDERS):
+            self.assertTrue(x in dir(turicreate))
+
+        # TODO Test whether or not things are NOT visible
+
+def load_lazy_modules():
```
Comment (Jarvi-Izana) [inline src/python/turicreate/test/test_api_visibility.py]: lazy_modules_force_load might be a better name. ;-)

Comment (abhishekpratapa) [reply src/python/turicreate/test/test_api_visibility.py]: I'll give this one to you.

**Comment by Jarvi-Izana** [comment]
great work! amazed!

File: src/python/turicreate/__init__.py (hunk 5 above)
Comment (abhishekpratapa) [reply src/python/turicreate/__init__.py]: Probably still leave it in for people to use if they want.

File: src/python/turicreate/test/test_api_visibility.py (hunk 6 above)
Comment (abhishekpratapa) [reply src/python/turicreate/test/test_api_visibility.py]: Done.

**Comment by TobyRoseman** [comment]
It's really difficult to see what has changed in the test file. Next time please use `git mv` to rename the test file.

**Comment by hoytak** [comment]
Can this be merged?

**Comment by Jarvi-Izana** [comment]
test failure due to the implicit import.

**Pull request closed without merging**